### PR TITLE
Issue 206 pt 4

### PR DESCRIPTION
**Pull request by MeghanBomberger**
For issue #206 

The dropdown, modal, and delete action are working however I am unsure how to tell it to refresh the client list after the deletion and am still getting these errors:

Unsure on this one entirely:
image

This one is because when the object key is deleted it doesn't exist. I've tried every way I can think of to make this read 'false' when needed with no luck. 
image

Closest I've gotten on this one is https://reactjs.org/docs/cross-origin-errors.html, but when I tried the recommended solution it crashed the site.
image

### CODE REVIEW

**Comment by MeghanBomberger** [comment]
Sorry about that, should be fixed now.

File: .gitignore [hunk 1]
```diff
@@ -68,4 +68,4 @@ static
 !.elasticbeanstalk/*.cfg.yml
 !.elasticbeanstalk/*.global.yml
 
-.env
```
Comment (MeghanBomberger) [inline .gitignore]: This edit was from an extension's file not being omitted as expected. I removed it from the project folder entirely and restored the .gitignore to it's previous state.

File: frontend/client-list/table/desktop-clients-table.component.tsx [hunk 2]
```diff
@@ -92,7 +92,8 @@ export default function DesktopClientsTable(props: ClientsTableProps) {
                   name="client-checked"
                   aria-label={`Select ${client.fullName}`}
                   value={client.id}
-                  onClick={evt => evt.stopPropagation()}
+                  checked={props.isSelected[client.id]}
```
Comment (MeghanBomberger) [inline frontend/client-list/table/desktop-clients-table.component.tsx]: This is the problem child line for the second error. 

Using a ternary, a separate function that returned true or false, and having the unclick change the value from true to false did not work. This way works on the surface, but still kicks out a typescript error.

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx [hunk 3]
```diff
@@ -122,4 +230,12 @@ type ClientsTableToolbarProps = {
   setPage(pageNum: number): void;
   setSearch(searchParse: SearchParseValues): any;
   fetchingClient: boolean;
+  selectedClients: any;
```
Comment (MeghanBomberger) [inline frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: I wasn't sure if this should be Array<object> or not

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx [hunk 4]
```diff
@@ -122,4 +230,12 @@ type ClientsTableToolbarProps = {
...
+  setSelectedClients: (client: object) => any;
+  modalIsOpen: boolean;
+  setModalIsOpen: any;
```
Comment (MeghanBomberger) [inline frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: Is using `any` here okay or should it be `(modalIsOpen: boolean) => any`

File: frontend/client-list/client-list.component.tsx [hunk 5]
```diff
@@ -39,7 +60,24 @@ export default function ClientList(props: ClientListProps) {
         newSortOrder={newSortOrder}
         sortField={apiState.sortField}
         sortOrder={apiState.sortOrder}
+        selectedClients={selectedClients}
+        setSelectedClients={setSelectedClients}
+        isSelected={isSelected}
+        setIsSelected={setIsSelected}
       />
+      {modalIsOpen ? (
```
Comment (joeldenning) [inline frontend/client-list/client-list.component.tsx]: nit:

```suggestion
      {modalIsOpen && (
```

File: frontend/client-list/table/clients-table.component.tsx [hunk 6]
```diff
@@ -25,4 +25,8 @@ export type ClientsTableProps = {
   newSortOrder: (sortField: SortField, sortOrder: SortOrder) => any;
   sortField: SortField;
   sortOrder: SortOrder;
+  selectedClients: Array<object>;
```
Comment (joeldenning) [inline frontend/client-list/table/clients-table.component.tsx]: nit: here and everywhere else we can use the ClientListClient type

File: frontend/client-list/table/desktop-clients-table.component.tsx [hunk 7]
```diff
@@ -181,6 +182,28 @@ export default function DesktopClientsTable(props: ClientsTableProps) {
...
+
+  function handleCheckBoxChange(id, data) {
+    const newIsSelected = props.isSelected;
+    const newSelection = props.selectedClients;
+    if (props.isSelected[id] === true) {
+      delete newIsSelected[id];
+      props.setIsSelected(newIsSelected);
+      const objectPosition = newSelection
+        .map(client => {
+          console.log(client);
```
Comment (joeldenning) [inline frontend/client-list/table/desktop-clients-table.component.tsx]: Should not push console.log's that are only there for debugging purposes.

File: frontend/client-list/table/desktop-clients-table.component.tsx [hunk 8]
```diff
@@ -181,6 +182,28 @@ export default function DesktopClientsTable(props: ClientsTableProps) {
...
     );
   }
+
+  function handleCheckBoxChange(id, data) {
+    const newIsSelected = props.isSelected;
+    const newSelection = props.selectedClients;
+    if (props.isSelected[id] === true) {
+      delete newIsSelected[id];
+      props.setIsSelected(newIsSelected);
+      const objectPosition = newSelection
```
Comment (joeldenning) [inline frontend/client-list/table/desktop-clients-table.component.tsx]: Easier way of doing this:
```js
props.setSelectedClients(props.setSelectedClients.filter(c => c.id !== props.id))
```

File: frontend/client-list/table/desktop-clients-table.component.tsx [hunk 9]
```diff
@@ -181,6 +182,28 @@ export default function DesktopClientsTable(props: ClientsTableProps) {
...
+        .map(client => {
+          console.log(client);
+          return client[id].id;
+        })
+        .indexOf(id);
+      newSelection.splice(objectPosition, 1);
+      props.setSelectedClients(newSelection);
+    } else {
+      newIsSelected[id] = true;
+      newSelection.push(data);
```
Comment (joeldenning) [inline frontend/client-list/table/desktop-clients-table.component.tsx]: It's best not to modify props by pushing data into the array. From the react docs:

> Remember that props are readonly. They should not be modified in any way:

https://reactjs.org/docs/glossary.html#props

Instead, you could do this:
```js
props.setSelectedClients(props.selectedClients.concat(data))
```

which does not modify the original array

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx [hunk 10]
```diff
@@ -1,21 +1,49 @@
...
       <div className="desktop-table-toolbar">
         <div className="left">
+          <div className="bulk-action-container">
+            <img
+              alt="kabob icon"
+              className="bulk-action-icon"
+              onClick={() => setBulkActionMenuIsOpen(!bulkActionMenuIsOpen)}
+              src={kabobIcon}
+            />
+            {bulkActionMenuIsOpen === true ? (
```
Comment (joeldenning) [inline frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: another example of `&&` being easier than ternaries.

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx [hunk 11]
```diff
@@ -71,6 +99,55 @@ export default function ClientsTableToolbar(props: ClientsTableToolbarProps) {
...
+        .catch(err => {
+          props.setModalIsOpen(false);
+          props.setModalOptions({});
+        });
+    }
+  }
+}
+
+function DeleteConfirmation(props) {
+  console.log(props.selectedClients);
```
Comment (joeldenning) [inline frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: remove console.logs for debugging.

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx [hunk 12]
```diff
@@ -90,6 +167,37 @@ const css = `
   align-items: center;
 }
 
+& .bulk-action-icon {
+  height: 2rem;
+  margin: 0 1rem 0 -0.5rem;
+}
+
+& .bulk-action-menu-modal {
```
Comment (joeldenning) [inline frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: Did the styleguide's popup class not work for you? Why did you recreate it? https://github.com/JustUtahCoders/comunidades-unidas-internal/blob/master/frontend/styleguide.component.tsx#L190

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx (hunk 3 above)
Comment (joeldenning) [reply frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: yeah `Array<ClientListClient>`

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx (hunk 4 above)
Comment (joeldenning) [reply frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: should be `setModalIsOpen: (isOpen: boolean)`

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx [hunk 13]
```diff
@@ -71,6 +99,55 @@ export default function ClientsTableToolbar(props: ClientsTableToolbarProps) {
...
+        primaryText: "No",
+        primaryAction: () => props.setModalIsOpen(false),
+        secondaryText: "Yes",
+        secondaryAction: () => deleteSelectedClients(),
+        children: <DeleteConfirmation selectedClients={props.selectedClients} />
+      });
+    }
+  }
+
+  function deleteSelectedClients() {
```
Comment (joeldenning) [inline frontend/client-list/toolbar/clients-table-toolbar.component.tsx]: i don't think this removes clients from the list after they are deleted.

File: .vscode/linenote/frontend/client-list/client-list.component.tsx#L96.md [hunk 14]
```diff
@@ -0,0 +1 @@
+Use as example for calling dispatchApiState ActionType.fetch to refresh after deletion
```
Comment (joeldenning) [inline .vscode/linenote/frontend/client-list/client-list.component.tsx#L96.md]: Please remove this file from source control and add it to the gitignore.

Add the following to the `.gitignore` file:
```
.vscode
```

Then run the following commands in a terminal:
```sh
git rm --cached .vscode
git status # the vscode files should now say it is going to be deleted
git add .
git commit
```

File: frontend/client-list/toolbar/clients-table-toolbar.component.tsx [hunk 15]
```diff
@@ -101,55 +99,102 @@ export default function ClientsTableToolbar(props: ClientsTableToolbarProps) {
...
+            headerText: null,
+            primaryText: null,
+            primaryAction: null,
+            secondaryText: null,
+            secondaryAction: null,
+            children: null
+          }),
+        children: (
+          <p>
+            You much select a client from the list before this action can be
```
Comment (joeldenning) [inline frontend/client-list/toolbar/clients-table-toolbar.component.tsx]:
```suggestion
            You must select a client from the list before this action can be
```

File: .vscode/linenote/frontend/client-list/client-list.component.tsx#L96.md (hunk 14 above)
Comment (MeghanBomberger) [reply .vscode/linenote/frontend/client-list/client-list.component.tsx#L96.md]: image

**Pull request merged**